src/Nethermind/Nethermind.Flashbots/Handlers/ValidateBuilderSubmissionHandler.cs

```diff
@@ -0,0 +1,427 @@
+    {
+        BlockHeader? parentHeader = _blockTree.FindHeader(block.ParentHash!, BlockTreeLookupOptions.DoNotCreateLevelIfMissing);
+
+        if (parentHeader is null)
+        {
+            error = $"Parent header {block.ParentHash} not found";
+            return false;
+        }
+
+        if (!ValidateBlockMetadata(block, registerGasLimit, parentHeader, out error))
+        {
+            return false;
+        }
+
+        using IReadOnlyTxProcessingScope processingScope = _readOnlyTxProcessingEnvFactory.Create().Build(parentHeader.StateRoot!);
+        IWorldState currentState = processingScope.WorldState;
+        ITransactionProcessor transactionProcessor = processingScope.TransactionProcessor;
+
+        UInt256 feeRecipientBalanceBefore = currentState.HasStateForRoot(currentState.StateRoot) ? (currentState.AccountExists(feeRecipient) ? currentState.GetBalance(feeRecipient) : UInt256.Zero) : UInt256.Zero;
+
+        IBlockCachePreWarmer preWarmer = new BlockCachePreWarmer(_readOnlyTxProcessingEnvFactory, _specProvider, _logManager);
+
+        BlockProcessor blockProcessor = CreateBlockProcessor(currentState, transactionProcessor, _flashbotsConfig.EnablePreWarmer ? preWarmer : null);
+
+        EthereumEcdsa ecdsa = new EthereumEcdsa(_specProvider.ChainId);
+        IReleaseSpec spec = _specProvider.GetSpec(parentHeader);
+
+        RecoverSenderAddress(block, ecdsa, spec);
+
+        List<Block> suggestedBlocks = [block];
+        BlockReceiptsTracer blockReceiptsTracer = new();
+
+        try
+        {
+            if (!_flashbotsConfig.EnableValidation)
+            {
+                ValidateSubmissionProcessingOptions |= ProcessingOptions.NoValidation;
+            }
+            Block processedBlock = blockProcessor.Process(currentState.StateRoot, suggestedBlocks, ValidateSubmissionProcessingOptions, blockReceiptsTracer)[0];
+        }
+        catch (Exception e)
+        {
+            error = $"Block processing failed: {e.Message}";
+            return false;
+        }
+
+        UInt256 feeRecipientBalanceAfter = currentState.GetBalance(feeRecipient);
+
+        UInt256 amtBeforeOrWithdrawn = feeRecipientBalanceBefore;
+
+        if (excludeWithdrawals)
+        {
+            foreach (Withdrawal withdrawal in block.Withdrawals ?? [])
+            {
+                if (withdrawal.Address == feeRecipient)
+                {
+                    amtBeforeOrWithdrawn += withdrawal.AmountInGwei;
+                }
+            }
+        }
+
+        if (!_blockValidator.ValidateSuggestedBlock(block, out error))
+        {
+            return false;
+        }
+
+        if (ValidateProposerPayment(expectedProfit, useBalanceDiffProfit, feeRecipientBalanceAfter, amtBeforeOrWithdrawn)) return true;
+
+        if (!ValidateProcessedBlock(block, feeRecipient, expectedProfit, [.. blockReceiptsTracer.TxReceipts], out error))
```

Comment:
This still allocates array, I thought of changing parameter type to `IReadOnlyList` for example.